File: benchmarks/livermore_canonical/k21_matmul.f
C     Kernel 21 -- Matrix*Matrix Product (canonical)
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C     Canonical body:
C       DO 21 k=1,25; DO 21 i=1,25; DO 21 j=1,n:
C         PX(i,j) = PX(i,j) + VY(i,k) * CX(k,j)
C
      PROGRAM K21
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      INTEGER REP
      DIMENSION PX(25,101), VY(101,25), CX(25,101)
C
C     Initialize PX to zero
C
      DO 6 J = 1, 101
         DO 5 I = 1, 25
            PX(I,J) = 0.0D0
    5    CONTINUE
    6 CONTINUE
C
C     Initialize VY using SIGNEL (column-major fill: outer K=1..25, inner I=1..101)
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 12 K = 1, 25
         DO 11 I = 1, 101
            BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
            FUZZ = -FUZZ
            VY(I,K) = (BUZZ - FIZZ) * 0.1D0
   11    CONTINUE
   12 CONTINUE
C
C     Initialize CX using SIGNEL (column-major fill: outer J=1..101, inner K=1..25)
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 14 J = 1, 101
         DO 13 K = 1, 25
            BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
            FUZZ = -FUZZ
            CX(K,J) = (BUZZ - FIZZ) * 0.1D0
   13    CONTINUE
   14 CONTINUE
C
      N = 101
C
C     Kernel loop (canonical: PX(i,j) += VY(i,k) * CX(k,j))
C
      DO 500 REP = 1, 100000
         DO 300 K = 1, 25
            DO 200 I = 1, 25
               DO 100 J = 1, N
                  PX(I,J) = PX(I,J) + VY(I,K) * CX(K,J)
  100          CONTINUE
  200       CONTINUE
  300    CONTINUE
  500 CONTINUE
C
      WRITE(*,900) PX(1,1)
  900 FORMAT('px(1,1) = ', E25.15)
      END

File: benchmarks/livermore_canonical/k21_matmul.c
/* K21 -- Matrix*Matrix Product (canonical, 1-based indexing)
   Canonical body:
     DO 21 k=1,25; DO 21 i=1,25; DO 21 j=1,n:
       PX(i,j) = PX(i,j) + VY(i,k) * CX(k,j)
   Column-major:
     PX dims (25,101): PX(i,j) at flat (j-1)*25 + i
     VY dims (101,25): VY(i,k) at flat (k-1)*101 + i
     CX dims (25,101): CX(k,j) at flat (j-1)*25 + k */
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define NJ    101
#define NK    25
#define NI    25
#define NREPS 100000

#define PX(i,j) px[((j)-1)*25  + (i)]
#define VY(i,k) vy[((k)-1)*101 + (i)]
#define CX(k,j) cx[((j)-1)*25  + (k)]

int main(void) {
    /* PX:25*101=2525, VY:101*25=2525, CX:25*101=2525 (all need [2526]) */
    double px[2526], vy[2526], cx[2526];

    for (long i = 1; i <= 25*101; i++) px[i] = 0.0;
    /* VY filled column-major: outer k=1..25, inner i=1..101 */
    signel(&vy[1], 101 * 25);
    /* CX filled column-major: outer j=1..101, inner k=1..25 */
    signel(&cx[1], 25 * 101);

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    for (long rep = 0; rep < NREPS; rep++) {
        for (long k = 1; k <= NK; k++) {
            for (long i = 1; i <= NI; i++) {
                for (long j = 1; j <= NJ; j++) {
                    PX(i,j) = PX(i,j) + VY(i,k) * CX(k,j);
                }
            }
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("px(1,1) = %f\n", PX(1,1));
    return 0;
}
